File: Leetcode/1038 -  Binary Search Tree to Greater Sum Tree.cpp
// https://leetcode.com/problems/binary-search-tree-to-greater-sum-tree/
// Tags - Binary Search Tree

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sum = 0;
    TreeNode* bstToGst(TreeNode* root) {
        if(root == NULL)
            return NULL;
        
        bstToGst(root->right);
        root->val = sum +=root->val;
        bstToGst(root->left);
        
        return root;
    }
};

File: 1038 - Binary Search Tree to Greater Sum Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def bstToGst(self, root: TreeNode) -> TreeNode:
    def traverse(node):
      if not node:
        return

      traverse(node.right)

      self.res += node.val
      node.val = self.res

      traverse(node.left)

    self.res = 0
    traverse(root)

    return root
